File: crack-data-structures-and-algorithms/leetcode/python-impl/meeting_rooms_II_q253.py
# 核心思路
# 两个数组分别保存所有的开始时间和结束时间，排序
# 计算有多少个区间可能重合的次数，等于需要会议室的数量
class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        if len(intervals) == 0:
            return 0

        stime = []
        etime = []
        for x in intervals:
            stime.append(x[0])
            etime.append(x[1])

        stime.sort()
        etime.sort()

        room_cnt = 0
        epos = 0
        for i in range(len(intervals)):
            if stime[i] < etime[epos]:
                room_cnt += 1
            else:
                # reuse the room, and update our endtime of room
                epos += 1

        return room_cnt


# 一个 suboptimal 但是很直接的解法
# rooms 保存已经开辟的会议室，如果一个会议室被复用，则延长结束时间
class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        if len(intervals) == 0:
            return 0

        intervals.sort()
        rooms = [intervals[0][:]]
        for i in range(1, len(intervals)):
            found = False
            for j in range(len(rooms)):
                if intervals[i][0] >= rooms[j][1]:
                    # extend room occupation time
                    rooms[j][1] = intervals[i][1]
                    found = True
                    break
            # If no room can be reused, make a new room.
            if not found:
                rooms.append(intervals[i][:])

        return len(rooms)

File: crack-data-structures-and-algorithms/leetcode/meeting_rooms_II_q253.cpp
#include <algorithm>
#include <vector>

using namespace std;

// 核心思路
// 将会议的开始时间/结束时间分别保存到st/et两个数组里，并且分别排序
// 对于 st 的每个开始时间，要么复用一个空的会议室，要么开一个新的会议室
// 而 et 数组维护了按时间排序的下一个可用会议室的开始时间，所以
// 1) 如果某个 stime < et[pos] 则说明目前没有可用会议室了，因为最接近的可用会议室要等到 et[pos]，需要新开会议室
// 2) 反之如果 stime >= et[pos] 则说明 et[pos] 时间开始可用的会议室可以被 stime 开始的会议复用，我们需要自增
//    pos 去检查下一个会议室的可用开始时间
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        vector<int> st;
        vector<int> et;
        for (const auto& pair : intervals) {
            st.push_back(pair[0]);
            et.push_back(pair[1]);
        }

        sort(st.begin(), st.end());
        sort(et.begin(), et.end());

        int room = 0;
        int epos = 0;
        for (int stime : st) {
            if (stime < et[epos]) {
                ++room;
            } else {
                ++epos;
            }
        }

        return room;
    }
};
